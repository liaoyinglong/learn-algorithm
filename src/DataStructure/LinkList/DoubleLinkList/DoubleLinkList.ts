/**
 * 双向链表
 */
export class DoubleLinkList<T> {
  private head?: Node<T>;
  private last?: Node<T>;
  private len = 0;

  get length() {
    return this.len;
  }

  add(item: T) {
    this.inset(item, this.len);
  }

  inset(item: T, targetIndex: number) {
    const pendingNode = new Node(item);
    if (!this.head) {
      this.head = pendingNode;
    } else if (!this.last) {
      this.last = pendingNode;
      this.head.next = this.last;
      this.last.prev = this.head;
    } else if (targetIndex === this.len) {
      this.last.next = pendingNode;
      pendingNode.prev = this.last;
      this.last = this.last.next;
    }
    this.len++;
  }
}

class Node<T> {
  next?: Node<T>;
  prev?: Node<T>;
  constructor(public data: T) {}
}
