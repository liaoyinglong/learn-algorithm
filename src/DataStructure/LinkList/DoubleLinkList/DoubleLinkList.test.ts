import { DoubleLinkList } from "./DoubleLinkList";

describe("SingleLinkList", () => {
  let list: DoubleLinkList<number | string>;

  beforeEach(() => {
    list = new DoubleLinkList<number | string>();
    list.add("a");
    list.add("b");
    list.add("c");
  });

  it("add should correctly", function () {
    expect(list).toMatchSnapshot();
    expect(list.length).toBe(3);
  });

  it("get should correctly", function () {
    expect(list.get(-1).data).toBe("a");
    expect(list.get(0).data).toBe("a");
    expect(list.get(1).data).toBe("b");
    expect(list.get(2).data).toBe("c");
    expect(list.get(22).data).toBe("c");
  });

  it("头插入 应该正确", function () {
    list.inset("0a", 0);
    expect(list).toMatchSnapshot();
  });

  it("指定index插入", function () {
    list.inset("1a", 1);
    expect(list).toMatchSnapshot();
    list.inset("0b", 2);
    expect(list).toMatchSnapshot();
  });
});
