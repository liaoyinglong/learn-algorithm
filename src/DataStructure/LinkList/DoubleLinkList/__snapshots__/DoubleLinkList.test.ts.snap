// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SingleLinkList add should correctly 1`] = `
DoubleLinkList {
  "head": Node {
    "data": "a",
    "next": Node {
      "data": "b",
      "next": Node {
        "data": "c",
        "prev": [Circular],
      },
      "prev": [Circular],
    },
  },
  "last": Node {
    "data": "c",
    "prev": Node {
      "data": "b",
      "next": [Circular],
      "prev": Node {
        "data": "a",
        "next": [Circular],
      },
    },
  },
  "len": 3,
}
`;
